File: my-folder/0200-number-of-islands/solution.java
class Solution {
    
  private int[][] offsets = {{0, -1}, {-1, 0}, {0, 1}, {1, 0}};
  
  public int numIslands(char[][] grid) {
    int count = 0;
    
    for (int i = 0; i < grid.length; i++) {
      for (int j = 0; j < grid[0].length; j++) {
        if (grid[i][j] == '1') {
          dfs(grid, i, j);
          count++;
        }
      }
    }
    
    return count;
  }
  
  private void dfs(char[][] grid, int row, int col) {
    grid[row][col] = '0';
    
    for (int[] offset : offsets) {
      int nRow = row + offset[0];
      int nCol = col + offset[1];
      
      if (nRow < 0 || nRow >= grid.length || nCol < 0 || nCol >= grid[0].length) {
        continue;
      }
      
      if (grid[nRow][nCol] == '1') {
        dfs(grid, nRow, nCol);
      }
    }
  }
}

File: my-folder/0200-number-of-islands/solution.swift
class Solution {
  
    var visited = Set<[Int]>()
  
    func numIslands(_ grid: [[Character]]) -> Int {
        
         var result = 0
      
         for row in 0..<grid.count{
          for col in 0..<grid[0].count{
               if grid[row][col] == "1" && !visited.contains([row, col]){
                 dfs(row, col, grid)
                 result += 1
               }
          }
         }
      
      return result
    }
  
     func dfs(_ row : Int, _ col : Int, _ grid : [[Character]]){
            if row < 0 || row >= grid.count || col < 0 || col >= grid[0].count || visited.contains([row, col]) || grid[row][col] == "0"{
                return 
            }
            
            visited.insert([row, col])
       
            dfs(row+1, col, grid)
            dfs(row-1, col, grid)
            dfs(row, col+1, grid)
            dfs(row, col-1, grid)
           
     }
}
